### ISSUE TITLE
Unable to reuse the ResulTypes

### ISSUE THREAD

**Issue by PuneetPunamiya** [goadesign/goa#2635]
I have created an example [here](https://github.com/PuneetPunamiya/goa-type-issue)

The basic gist of the issue is :

Consider an example:

```code
var first = ("application/vnd.hub.first", func() {


  Attribute("id", UInt, "ID is the unique id of the fruit", func() {
		Example("id", 1)
	})
  Attribute("name", String, "Name of Fruit", func() {
		Example("name", "apple")
	})
  Attribute("second", ArrayOf("second"))    <------ cannot use ResultType `second` here without an array as well
  
})
```

```code
var second = (application/vnd.hub.second, func() {


  Attribute("id", UInt, "ID is the unique id of the fruit", func() {
		Example("id", 2)
	})
  Attribute("name", String, "Name of Fruit", func() {
		Example("name", "mango")
	})
  Attribute("first", ArrayOf("first"))
  
})
```

- It fails to create the `gen` directory

  - The error is :

    ```bash
    $ goa gen goa-type-issue/design 

    exit status 1
    [error] invalid ArrayOf argument: not a type and not a known user type name in attribute

    ```

**Comment by raphael** [comment]
PR #2644 makes it possible to define the type name directly as a parameter of the `ResultType` function. This ensures that the name is set prior to being referenced in a different DSL function. So in your case you can now use:
```go
var first = ResultType("application/vnd.hub.first", "first", func() {
    Attribute("id", UInt, "ID is the unique id of the fruit", func() {
        Example("id", 1)
    })
    Attribute("name", String, "Name of Fruit", func() {
        Example("name", "apple")
    })
    Attribute("second", ArrayOf("second"))  
})

var second = ResultType("application/vnd.hub.second", "second", func() {
    Attribute("id", UInt, "ID is the unique id of the fruit", func() {
        Example("id", 2)
    })
    Attribute("name", String, "Name of Fruit", func() {
        Example("name", "mango")
    })
    Attribute("first", ArrayOf("first"))
})
```

**Comment by raphael** [comment]
Thank you for the report, this is fixed in v3.2.1.
